x/evm/types/state_transition.go

```diff
@@ -109,6 +109,28 @@ func (st StateTransition) newEVM(
 	return vm.NewEVM(blockCtx, txCtx, csdb, config.EthereumConfig(st.ChainID), vmConfig)
 }
 
+func (st StateTransition) refundGas(gasInfo GasInfo) {
+	// Apply refund counter, capped to half of the used gas.
+	refund := gasInfo.GasConsumed / 2
+	if refund > gasInfo.GasRefunded {
+		refund = gasInfo.GasRefunded
+	}
+
+	// Return ETH for remaining gas, exchanged at the original rate.
+	gasRefund := new(big.Int).Mul(new(big.Int).SetUint64(refund), st.Price)
+
+	// add refund back to counter.
+	// add gas information into state db
+	st.Csdb.AddRefund(gasRefund.Uint64())
+	st.Csdb.AddBalance(st.Sender, gasRefund)
+
+	// Also return remaining gas to the block gas counter so it is
+	// available for the next transaction.
+	// st.gp.AddGas(st.gas)
```

Comment:
as I remember you mentioned, the gas counter is a cumulative value that only adds up consumed gas in each tx, so that we don't need to recap it after. is it the reason you commented this part?

Reply:
this piece of code was taken from geth directly (federico had added part here)
commented out because it's just not correct on ethermint but was contemplating if we needed similar behaviour